modules/nlu/src/backend/pipelines/slots/pre-processor.ts

```diff
@@ -78,36 +106,34 @@ export const generateTrainingSequence = (langProvider: LanguageProvider) => asyn
   slotDefinitions: sdk.NLU.SlotDefinition[],
   intentName: string = '',
   contexts: string[] = []
-): Promise<Sequence> => {
+): Promise<TrainingSequence> => {
   let tokens: Token[] = []
-  let matches: RegExpExecArray | null
   const genToken = _generateTrainingTokens(langProvider)
   const cannonical = keepEntityValues(input)
-  let inputCopy = input
+  const knownSlots = getKnownSlots(input, slotDefinitions)
 
-  do {
-    matches = ITTERATIVE_SLOTS_REGEX.exec(inputCopy)
+  // TODO: this logic belongs near makeTokens and we should let makeTokens fill the matched entities
+  for (const slot of knownSlots) {
+    const start = _.isEmpty(tokens) ? 0 : _.last(tokens)!.end
+    const sub = cannonical.substring(start, slot.start - 1)
+    const tokensBeforeSlot = await genToken(sub, lang, start)
 
-    if (matches) {
-      const sub = inputCopy.substr(0, matches.index - 1)
-      const start = _.isEmpty(tokens) ? 0 : _.last(tokens)!.end
-      const tokensBeforeSlot = await genToken(sub, lang, start)
+    const slotTokens = await genToken(slot.source, lang, slot.start, slot.name, slotDefinitions)
 
-      const slotTokens = await genToken(matches[1], lang, start + matches.index, matches[2], slotDefinitions)
-
-      tokens = [...tokens, ...tokensBeforeSlot, ...slotTokens]
-      inputCopy = inputCopy.substr(matches.index + matches[0].length).trim()
-    }
-  } while (matches)
+    tokens = [...tokens, ...tokensBeforeSlot, ...slotTokens]
+  }
 
-  if (inputCopy.length) {
-    tokens = [...tokens, ...(await genToken(inputCopy, lang, _.isEmpty(tokens) ? 0 : _.last(tokens)!.end))]
+  const lastSlot = _.maxBy(knownSlots, ks => ks.end)
+  if (lastSlot && lastSlot!.end < cannonical.length) {
```

Comment:
can't wrap my head around this, can you add a comment that explains what that does?

Reply:
Sure, it only means: if lastSlot is not the end of input.

Ex:

var cannonical = "some text... slot#1 some other text slot#2 finally some text..."
                                                <---->                           <---->

Here "slot#2" is not the end of the hole input

Exclamation mark (!) is just a way of saying to the compiler that lastSlot is defined.

https://stackoverflow.com/questions/42273853/in-typescript-what-is-the-exclamation-mark-bang-operator-when-dereferenci